PhysicsTools/NanoAOD/plugins/HoverEValueMapProducer.cc

```diff
@@ -0,0 +1,162 @@
+
+ Implementation:
+     [Notes on implementation]
+*/
+//
+//          Author:  Shubham Dutta
+//         Created:  Tue, 01 Nov 2022 07:45 IST
+//
+//
+
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "CommonTools/Egamma/interface/EffectiveAreas.h"
+
+#include "DataFormats/PatCandidates/interface/Muon.h"
+#include "DataFormats/PatCandidates/interface/Electron.h"
+#include "DataFormats/PatCandidates/interface/Photon.h"
+#include "DataFormats/PatCandidates/interface/IsolatedTrack.h"
+
+//
+// class declaration
+//
+
+template <typename T>
+class HoverEValueMapProducer : public edm::global::EDProducer<> {
+public:
+  explicit HoverEValueMapProducer(const edm::ParameterSet& iConfig)
+      : src_(consumes<edm::View<T>>(iConfig.getParameter<edm::InputTag>("src"))),
+        relative_(iConfig.getParameter<bool>("relative")) {
+    if ((typeid(T) == typeid(pat::Photon))) {
+      produces<edm::ValueMap<float>>("HoEForPhoEACorr");
+
+      rho_ = consumes<double>(iConfig.getParameter<edm::InputTag>("rho"));
+
+      quadratic_ea_hOverE_ = std::make_unique<EffectiveAreas>(
+          (iConfig.getParameter<edm::FileInPath>("QuadraticEAFile_HoverE")).fullPath(), true);
+    }
+  }
+  ~HoverEValueMapProducer() override {}
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+  // ----------member data ---------------------------
+
+  edm::EDGetTokenT<edm::View<T>> src_;
+  bool relative_;
+  edm::EDGetTokenT<double> rho_;
+  std::unique_ptr<EffectiveAreas> quadratic_ea_hOverE_;
+  float getEtaForEA(const T*) const;
+  void doHoverEPho(edm::Event&) const;
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+template <typename T>
+float HoverEValueMapProducer<T>::getEtaForEA(const T* obj) const {
+  return obj->eta();
+}
+
+template <>
+float HoverEValueMapProducer<pat::Photon>::getEtaForEA(const pat::Photon* ph) const {
+  return ph->superCluster()->eta();
+}
+
+template <typename T>
+void HoverEValueMapProducer<T>::produce(edm::StreamID streamID,
+                                        edm::Event& iEvent,
+                                        const edm::EventSetup& iSetup) const {
+  if ((typeid(T) == typeid(pat::Photon))) {
+    doHoverEPho(iEvent);
+  }
+}
+
+template <typename T>
+void HoverEValueMapProducer<T>::doHoverEPho(edm::Event& iEvent) const {}
+
+template <>
+void HoverEValueMapProducer<pat::Photon>::doHoverEPho(edm::Event& iEvent) const {
+  edm::Handle<edm::View<pat::Photon>> src;
```

Comment:
Please use the new style:
```
auto src = iEvent.getHandle(src_);
const double rho = iEvent.get(rho_);
```

Reply:
Ok, thanks. I will make this change and update the branch.